lib/DB/Schema/ResultSet/Course.pm

```diff
@@ -262,36 +335,202 @@ This gets the Course Settings for a course
 
 =over
 
-=item * hashref containing info about the course
+=item * C<info>, hashref containing info about the course
+
+=item * C<merged>, a boolean on whether the course setting is merged with its corresponding
+global setting.
 
 =item * C<$as_result_set>, a boolean if the return is to be a result_set
 
 =back
 
 =head3 output
 
-An array of courses as a C<DBIx::Class::ResultSet::Course> object
+An array of course settings as a C<DBIx::Class::ResultSet::CourseSetting> object
 if C<$as_result_set> is true.  Otherwise an array of hash_ref.
 
 =cut
 
 sub getCourseSettings ($self, %args) {
-	my $course = $self->getCourse(info => $args{info}, as_result_set => 1);
+	my $course           = $self->getCourse(info => $args{info}, as_result_set => 1);
+	my @settings_from_db = $course->course_settings;
+
+	return \@settings_from_db if $args{as_result_set};
+	my @settings_to_return = ($args{merged})
+		? map {
+			{ $_->get_inflated_columns, $_->global_setting->get_inflated_columns };
+		} @settings_from_db
+		: map {
+			{ $_->get_inflated_columns };
+		} @settings_from_db;
+
+	for my $setting (@settings_to_return) {
+		# value and default_value are decoded from JSON as a hash.  Return to a value.
+		for my $key (qw/default_value value/) {
+			$setting->{$key} = $setting->{$key}->{value} if defined($setting->{$key});
+		}
+	}
```

Comment:
Again, we're using two separate loops over the same data.

Also, it's unnecessary to use `->` more than once when traversing nested hashes.

Finally, what is going on in line 370? Replace a key-value if the key already has a defined value? Shouldn't this be `if defined($setting->{$key}{value})`?

Reply:
The inflator cleans this up a lot.